extension/js/common/core/crypto/smime/smime-key.ts

```diff
@@ -84,14 +91,26 @@ export class SmimeKey {
     return { data: new Uint8Array(arr), type: 'smime' };
   }
 
-  private static parsePemCertificate = (text: string): Key => {
-    const certificate = forge.pki.certificateFromPem(text);
-    SmimeKey.removeWeakKeys(certificate);
-    const email = (certificate.subject.getField('CN') as { value: string }).value;
-    const normalizedEmail = Str.parseEmail(email).email;
-    if (!normalizedEmail) {
-      throw new UnreportableError(`This S/MIME x.509 certificate has an invalid recipient email: ${email}`);
+  private static getNormalizedEmailsFromCertificate = (certificate: forge.pki.Certificate): string[] => {
+    const emailFromSubject = (certificate.subject.getField('CN') as { value: string }).value;
+    const normalizedEmail = Str.parseEmail(emailFromSubject).email;
+    const emails = normalizedEmail ? [normalizedEmail] : [];
+    // search for e-mails in subjectAltName extension
+    const subjectAltName = certificate.getExtension('subjectAltName') as { altNames: { type: number, value: string }[] };
+    if (subjectAltName && subjectAltName.altNames) {
+      const emailsFromAltNames = subjectAltName.altNames.filter(entry => entry.type === 1).
+        map(entry => Str.parseEmail(entry.value).email).filter(Boolean);
+      emails.push(...emailsFromAltNames as string[]);
+    }
+    if (emails.length) {
+      return emails.filter((value, index, self) => self.indexOf(value) === index);
     }
+    throw new UnreportableError(`This S/MIME x.509 certificate has an invalid recipient email: ${emailFromSubject}`);
+  }
+
+  private static getKeyFromCertificate = (certificate: forge.pki.Certificate, pem: string): Key => {
+    SmimeKey.removeWeakKeys(certificate);
+    const emails = SmimeKey.getNormalizedEmailsFromCertificate(certificate);
     const key = {
       type: 'x509',
       id: certificate.serialNumber.toUpperCase(),
```

Comment:
I'm afraid we can't use `serialNumber` as `fingerprint`.
Should we open a new task for this as well as include migration to re-calculate ids?

Reply:
Created an issue to discuss